### PR TITLE
Bulk Services / Bed Nights v2

### PR DESCRIPTION
**Pull request by gigxz**
## V2 Bulk Services & Bed Nights

Depends on hmis-warehouse PR: https://github.com/greenriver/hmis-warehouse/pull/4087

## How to Test
* BED NIGHTS: go to a night-by-night shelter project page
* BULK SERVICES: mark one or more CUstomServiceTypes as "supports_bulk_assignment" in the database, and go to the project page for a project that supports that service
* COC SELECTION: add multiple ProjectCoC records to the project
* UNIT ASSIGNMENT: add units to the project. when you enroll someone, they should be assigned to the next available unit
see also [qa instructions](https://www.pivotaltracker.com/story/show/187011913/comments/240472290)

## Type of change
- [x] New feature (adds functionality)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (eslint)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: src/hooks/useSearchParamState.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
+import { isDate } from 'date-fns';
+import { isNil } from 'lodash-es';
+import { useSearchParams } from 'react-router-dom';
+import { formatDateForGql, parseHmisDateString } from '@/modules/hmis/hmisUtil';
+
+// adapted from https://github.com/jschwindt/react-use-search-params-state/tree/main
```
Comment (gigxz) [inline src/hooks/useSearchParamState.ts:7]: check in -- should we add this as an npm package to give proper credit

File: src/modules/dataFetching/components/GenericTableWithData.tsx [hunk 2]
```diff
@@ -360,14 +364,17 @@ const GenericTableWithData = <
       )}
       <Box sx={containerSx}>
         <GenericTable<RowDataType>
-          loading={loading && !data}
+          // if loadingVariant was passed explicitly, show the loading state for refetches.
+          // otherwise only show the initial loading state
+          loading={loadingVariant ? loading : loading && !data}
+          loadingVariant={loadingVariant}
```
Comment (gigxz) [inline src/modules/dataFetching/components/GenericTableWithData.tsx]: janky, revisit

File: src/modules/services/components/bulk/AssignServiceButton.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  const { bulkAssign, bulkRemove, loading, apolloError } =
+    useBulkAssignMutations();
+
+  // Locally disable the button while the table refetches, AFTER this specific mutation has completed
+  const [localDisabled, setLocalDisabled] = useState(false);
+
+  // When the entire table has finished re-fetching, remove the loading state for this button.
+  useEffect(() => {
+    if (!tableLoading) setLocalDisabled(false);
+  }, [tableLoading]);
```
Comment (gigxz) [inline src/modules/services/components/bulk/AssignServiceButton.tsx:38]: Some funkiness here. `useBulkAssignMutations` uses apollo `refetchQueries` to specify that the table query should always refetch after a mutation completes. During that refetch, we want to keep this button disabled.

File: src/routes/routeUtil.ts [hunk 4]
```diff
@@ -15,3 +16,39 @@ export const locationFromDefaultOrLogin = (location: Location): boolean => {
...
   return !!location.state?.fromLoginRedirect;
 };
+
+export const PREV_SEARCH_KEY = 'prev';
+export const prevSearchParam = () =>
+  `${PREV_SEARCH_KEY}=${encodeURIComponent(
+    window.location.pathname + window.location.search
+  )}`;
+
+export const backToFromSearchParam = (
```
Comment (gigxz) [inline src/routes/routeUtil.ts]: 🚨 this seems to be working great, but just want to flag it is new and kinda complicated. Constructs a `To` based on the `prev` search param if its present, with the ability to inject search params.

**Comment by gigxz** [comment]
Alright @ttoomey @martha  - ready for review. I apologize for the enormity! I tried to call out some interesting parts in the comments.

This is up on QA (in a few minutes) as well.

( tagging @martha as secondary QA, so please don't feel the need to go through everything!)

File: src/api/operations/bedNight.queries.graphql [hunk 5]
```diff
@@ -1,52 +0,0 @@
-query GetProjectEnrollmentsForBedNights(
```
Comment (martha) [inline src/api/operations/bedNight.queries.graphql:1]: Is this being deleted because bed nights will only be used through bulk services from now on?

File: src/hooks/useSearchParamState.ts (hunk 1 above)
Comment (martha) [reply src/hooks/useSearchParamState.ts:7]: I guess the pros are giving proper credit and being able to benefit from improvements to that repo. But the con is adding a dependency to something that seems to be maintained by just one person, and isn't very widely used. Could this be a security concern as well?

We could look at whether there are other npm packages that are more popular and by extension maybe more trustworthy, maybe [this](https://www.npmjs.com/package/use-query-params?activeTab=readme) provides similar functionality?

File: src/modules/services/components/bulk/AssignServiceButton.tsx (hunk 3 above)
Comment (martha) [reply src/modules/services/components/bulk/AssignServiceButton.tsx:38]: This seems sensible. Maybe it would be good to add what you've written here in the github PR comment as a comment in the code as well?

File: src/routes/routeUtil.ts (hunk 4 above)
Comment (martha) [reply src/routes/routeUtil.ts]: This is cool! I would just say that the name `backToFromSearchParams` adds some cognitive load, haha! Even after understanding that `To` is the type of the object being returned, I had to concentrate to parse this, I think due to the `to` and `from` words right next to each other in the name. Also I wasn't sure `back` is needed in the name, maybe it's a synonym for `prev` and we should use that instead for consistency? maybe `prevUrlFromSearchParam` is the clearest, but on the other hand, that doesn't indicate that you are returning a type `To` instead of a type `URL`. Maybe `toPrevUrlFromSearchParam` although that's getting verboser.

Either way I also think it would be good to add a comment similar to what you've written here in gh, explaining what this is doing since the name is likely to end up a bit hard to parse either way

**Review by martha — COMMENTED** [review]
Amazing ‼️

File: src/api/operations/bedNight.queries.graphql (hunk 5 above)
Comment (gigxz) [reply src/api/operations/bedNight.queries.graphql:1]: Yep

File: src/modules/services/components/bulk/BulkServicesTable.tsx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        defaultFilters={defaultFilters}
+        // TODO: add user-facing filter options for enrolled clients and bed night date. No filter options for now.
+        filters={(f) => omit(f, 'project', 'organization')}
+        filterInputType='ClientFilterOptions'
+        defaultSortOption={
+          searchTerm
+            ? ClientSortOption.BestMatch
+            : ClientSortOption.LastNameAToZ
+        }
+        EnhancedTableToolbarProps={{
```
Comment (ttoomey) [inline src/modules/services/components/bulk/BulkServicesTable.tsx:189]: When you've scrolled into the table this top header isn't visible. Then when selecting a checkbox it's not clear why all the row buttons are disabled. What if we made this bar stick to the top (or bottom) so it's always visible to the user? Or maybe animate-in even?

File: src/routes/routeUtil.ts (hunk 4 above)
Comment (ttoomey) [reply src/routes/routeUtil.ts]: I think this is totally fine for now but I'd want to think hard about adopting this approach elsewhere. 

I think it's great to put navigational, filter, and pagination params into the URL; it's a better user experience imo. However I'm a little unsure about using the URL for workflow state that is internal / ephemeral.

An alternative might be something like react router location state that wouldn't expose this to the user:

```
navigate({pathname: bednights}, {state: })
```

File: src/modules/services/types.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { sub } from 'date-fns';
+
+export type ServicePeriod = {
+  start: Date;
+  end: Date;
+};
+
+const today = new Date();
```
Comment (ttoomey) [inline src/modules/services/types.ts:8]: I think there's a bug here. If the day rolls over while the app is open, today will not update. Not sure how much of a problem it would be but it might impact staff working late at a shelter. This probably wants to be a hook which would let the date at least update when components mounts. Ideally it would be a useCurrentDate hook.

maybe something like this (not tested):
```ts
import { endOfDay, differenceInMilliseconds } from 'date-fns';
import { useState, useEffect } from 'react';

const timeUntilMidnight = () => {
  const now = new Date();
  const midnight = endOfDay(now);
  return differenceInMilliseconds(midnight, now);
};

const useCurrentDate = () => {
  const [value, setValue] = useState(new Date());

  useEffect(() => {
    let timeout = setTimeout(function tick() {
      setValue(new Date());
      timeout = setTimeout(tick, timeUntilMidnight());
    }, timeUntilMidnight());

    return () => clearTimeout(timeout);
  }, []);

  return value;
};

export default useCurrentDate;
```

File: src/config/theme.ts [hunk 8]
```diff
@@ -250,6 +250,7 @@ const createThemeOptions = (theme: Theme) => ({
       },
       styleOverrides: {
         root: {
+          whiteSpace: 'nowrap',
```
Comment (ttoomey) [inline src/config/theme.ts]: This means button text won't wrap, right? Is this to handle wide buttons on mobile? I wonder if it would better to allow the text to wrap when needed and adjust the container to better manage the layout. For example, using flex-wrap:wrap on the container and flex-shrink:0 on the buttons.

If button text can't wrap, you could have buttons that overflow the container causing weird horizontal scroll.

File: src/hooks/useSearchParamState.ts (hunk 1 above)
Comment (ttoomey) [reply src/hooks/useSearchParamState.ts:7]: So the intention is to centralize casting query params from strings. It also provides a properly typed setter for params. That seems reasonable.

I expect we will need to modify this code, for example adding encrypted fields. So it's reasonable to keep it here for us to hack on.

The use-query-params lib is more widely used but it looks like it may not have an active maintainer (1 yr since last commit).

File: src/hooks/useSearchParamState.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  });
+  return allUrlParams;
+};
+
+const useSearchParamsState = (paramsDefinition: SearchParamsStateType) => {
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const values = getValues(paramsDefinition, searchParams);
+
+  const setValues = (newValues: Record<string, any>) => {
```
Comment (ttoomey) [inline src/hooks/useSearchParamState.ts]: memoize?
```suggestion
  const setValues = useCallback((newValues: Record<string, any>) => {
```

File: src/hooks/useSearchParamState.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      allUrlParams[key] = value;
+    }
+  });
+  return allUrlParams;
+};
+
+const useSearchParamsState = (paramsDefinition: SearchParamsStateType) => {
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const values = getValues(paramsDefinition, searchParams);
```
Comment (ttoomey) [inline src/hooks/useSearchParamState.ts]: memo?
```suggestion
  const values = useMemo(() => getValues(paramsDefinition, searchParams), [paramsDefinition, searchParams]);
```

File: src/modules/services/components/bulk/BulkServicesPage.tsx [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    );
+  }, [project, navigate, serviceTypeIdProp, setFilterParams]);
+
+  const stepThreeTitle = `Assign ${pluralize(serviceTypeName, 2)}`;
+
+  return (
+    <>
+      <PageTitle title={title} />
+      <Grid container rowSpacing={1}>
+        <Grid item sm={12} md={8} lg={8} xl={4}>
```
Comment (ttoomey) [inline src/modules/services/components/bulk/BulkServicesPage.tsx]: Do we need to encapsulate the steps in a Grid? This is subjective but I think it looks strange with the step cards being different widths. Maybe simpler to let them all be 100% width, take out the grid, and just throw margin bottom on each card? Cuts some of the indentation out this component too, which would be nice

File: src/modules/services/components/bulk/BulkServicesPage.tsx [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+                    setFilterParams({ serviceTypeId: option?.code })
+                  }
+                  label='Service Type'
+                  bulk
+                />
+              )}
+              <DatePicker
+                value={serviceDate}
+                onChange={(date) => setFilterParams({ serviceDate: date })}
+                max={new Date()}
```
Comment (ttoomey) [inline src/modules/services/components/bulk/BulkServicesPage.tsx:162]: might be another place for a hypothetical useCurrentDate()

File: src/hooks/useSearchParamState.ts [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        allUrlParams[key] = [allUrlParams[key], value];
+      }
+    } else {
+      allUrlParams[key] = value;
+    }
+  });
+  return allUrlParams;
+};
+
+const useSearchParamsState = (paramsDefinition: SearchParamsStateType) => {
```
Comment (ttoomey) [inline src/hooks/useSearchParamState.ts:108]: right now, the values returned from this hook are `any`. If we are going to use this utility more broadly, we should probably fix up the types to match the paramsDefinition using generics. Something for a future pass

File: src/modules/services/components/bulk/ServiceDateRangeSelect.tsx [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+          renderOption={renderOption}
+          disableClearable
+        />
+        <Button
+          color='error'
+          variant='outlined'
+          onClick={handleClear}
+          disabled={disabled || !selected}
+        >
+          Clear Results
```
Comment (ttoomey) [inline src/modules/services/components/bulk/ServiceDateRangeSelect.tsx]: I think something like "Reset Search" is more clear

File: src/modules/services/components/bulk/BulkServicesPage.tsx [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+                  onSearch={(value) => setFilterParams({ searchTerm: value })}
+                  onClearSearch={() => setFilterParams({ searchTerm: null })}
+                  label={null}
+                  placeholder='Client Name, DOB, SSN or ID'
+                  helperText='Search includes all of HMIS'
+                  ClearButtonProps={{
+                    color: 'error',
+                    startIcon: null,
+                    variant: 'outlined',
+                    children: 'Clear Results',
```
Comment (ttoomey) [inline src/modules/services/components/bulk/BulkServicesPage.tsx]: There may be no results found so I think something like "Reset Search" is more clear

File: src/modules/services/components/bulk/AssignServiceButton.tsx (hunk 3 above)
Comment (ttoomey) [reply src/modules/services/components/bulk/AssignServiceButton.tsx:38]: would it work to just disable the buttons when tableLoading is true?
```jsx
<LoadingButton disabled={disabled || localDisabled || tableLoading} ... />
```

**Review by ttoomey — COMMENTED** [review]
This is awesome! This is a huge piece of work and it's come together really well! 🙌 

I left a few suggestions and ideas but there's nothing that should be a merge blocker

File: src/modules/services/components/bulk/BulkServicesTable.tsx (hunk 6 above)
Comment (gigxz) [reply src/modules/services/components/bulk/BulkServicesTable.tsx:189]: I noticed the same issue. I will add a tooltip to the disabled buttons so at least you can get a message there.
[commented in figma](https://www.figma.com/file/F9EHR2IZYuMAlOJ9aOG2IL?node-id=2088:66116&mode=design#738730183), we can add a separate story.

File: src/modules/services/types.ts (hunk 7 above)
Comment (gigxz) [reply src/modules/services/types.ts:8]: Good catch, thanks

File: src/config/theme.ts (hunk 8 above)
Comment (gigxz) [reply src/config/theme.ts]: Removed this change

File: src/modules/services/types.ts (hunk 7 above)
Comment (gigxz) [reply src/modules/services/types.ts:8]: Although, I don't know. I don't think we want that behavior for this page. If they reach midnight while they are using the application to record bednights, they are still going to want to record it for the same date.. you don't want to record bed nights for 3/14 at 1am on 3/14 - you would still be recording for 3/13 at that point. See what I mean?

File: src/modules/services/components/bulk/ServiceDateRangeSelect.tsx (hunk 14 above)
Comment (gigxz) [reply src/modules/services/components/bulk/ServiceDateRangeSelect.tsx]: I agree

File: src/modules/services/components/bulk/AssignServiceButton.tsx (hunk 3 above)
Comment (gigxz) [reply src/modules/services/components/bulk/AssignServiceButton.tsx:38]: > would it work to just disable the buttons when tableLoading is true? 

that would disabled all the buttons in the table. we only want to disable the buttons that just performed a mutation

File: src/modules/services/components/bulk/BulkServicesPage.tsx (hunk 11 above)
Comment (gigxz) [reply src/modules/services/components/bulk/BulkServicesPage.tsx]: Design has staggered cards like this. But I feel like the complexity isnt really worth it.. removed

File: src/routes/routeUtil.ts (hunk 4 above)
Comment (gigxz) [reply src/routes/routeUtil.ts]: Cool,  thanks both. I updated to use react-router-dom state for the household navigation piece: https://github.com/greenriver/hmis-frontend/pull/676/commits/ec0e90833ef288efaaec4c0ba295cf5e1af64f9e

File: src/hooks/useSearchParamState.ts (hunk 1 above)
Comment (gigxz) [reply src/hooks/useSearchParamState.ts:7]: leaving this as is. per theron's other comment, too, if we use it going forward we need to make it type safe. [ticket for that](https://www.pivotaltracker.com/story/show/187258786)

File: src/modules/services/types.ts (hunk 7 above)
Comment (ttoomey) [reply src/modules/services/types.ts:8]: > If they reach midnight while they are using the application to record bednights, they are still going to want to record it for the same date

That makes sense and it should be fine as-is for now. However I think that it would be preferable to implement the logic explicitly in the component with an offset (the default date rolls over at 6am or something). That means if you reload the page after midnight, you'd get yesterday's date as desired.

Comment (gigxz) [reply src/modules/services/types.ts:8]: 👍  https://www.pivotaltracker.com/story/show/187263616
